controller/konnect/reconciler_generic.go

```diff
@@ -436,11 +524,184 @@ func getAPIAuthRefNN[T SupportedKonnectEntityType, TEnt EntityType[T]](
 	)
 }
 
+func getServiceRef[T SupportedKonnectEntityType, TEnt EntityType[T]](
+	e TEnt,
+) mo.Option[configurationv1alpha1.ServiceRef] {
+	switch e := any(e).(type) {
+	case *configurationv1alpha1.KongService,
+		*configurationv1.KongConsumer,
+		*konnectv1alpha1.KonnectControlPlane:
+		return mo.None[configurationv1alpha1.ServiceRef]()
+	case *configurationv1alpha1.KongRoute:
+		if e.Spec.ServiceRef == nil {
+			return mo.None[configurationv1alpha1.ServiceRef]()
+		}
+		return mo.Some(*e.Spec.ServiceRef)
+	default:
+		panic(fmt.Sprintf("unsupported entity type %T", e))
+	}
+}
+
+// handlKongServiceRef handles the ServiceRef for the given entity.
+// It sets the owner reference to the referenced KongService and updates the
+// status of the entity based on the referenced KongService status.
+func handlKongServiceRef[T SupportedKonnectEntityType, TEnt EntityType[T]](
+	ctx context.Context,
+	cl client.Client,
+	ent TEnt,
+) (ctrl.Result, error) {
+	kongServiceRef, ok := getServiceRef(ent).Get()
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+	switch kongServiceRef.Type {
+	case configurationv1alpha1.ServiceRefNamespacedRef:
+		svc := configurationv1alpha1.KongService{}
+		nn := types.NamespacedName{
+			Name: kongServiceRef.NamespacedRef.Name,
+			// TODO: handle cross namespace refs
+			Namespace: ent.GetNamespace(),
+		}
+
+		if err := cl.Get(ctx, nn, &svc); err != nil {
+			if res, errStatus := updateStatusWithCondition(
+				ctx, cl, ent,
+				KongServiceRefValidConditionType,
+				metav1.ConditionFalse,
+				KongServiceRefReasonInvalid,
+				err.Error(),
+			); errStatus != nil || res.Requeue {
+				return res, errStatus
+			}
+
+			return ctrl.Result{}, fmt.Errorf("Can't get the referenced KongService %s: %w", nn, err)
+		}
+
+		// If referenced KongService is being deleted, return an error so that we
+		// can remove the entity from Konnect first.
+		if delTimestamp := svc.GetDeletionTimestamp(); !delTimestamp.IsZero() {
+			return ctrl.Result{}, ReferencedKongServiceIsBeingDeleted{
+				Reference: nn,
+			}
+		}
+
+		cond, ok := k8sutils.GetCondition(KonnectEntityProgrammedConditionType, &svc)
+		if !ok || cond.Status != metav1.ConditionTrue /*|| cond.ObservedGeneration != cp.GetGeneration() */ {
```

Comment:
Why do we comment this condition out, could we explain it?

Reply:
Commented unnecessarily. Removed. PTAL